File: java/top/abosen/leetcode/ValidParentheses.java
package top.abosen.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;

//[20] 有效的括号
/* "()" */
public class ValidParentheses {

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public boolean isValid(String s) {
            Deque<Character> deque = new ArrayDeque<>();
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (c == '(' || c == '[' || c == '{') {
                    deque.push(c);
                } else {
                    if (deque.isEmpty()) return false;
                    char p = deque.poll();
                    if ((p != '(' || c != ')') && (p != '[' || c != ']') && (p != '{' || c != '}')) {
                        return false;
                    }
                }
            }
            return deque.isEmpty();
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/ValidParentheses.kt
package leetcode.editor.cn

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import java.util.*

// 有效的括号
// 20
/* "()" */

fun main() {
    val result = ValidParentheses.Solution()::isValid
    assertThat(result("()"), equalTo(true))
    assertThat(result("()[]{}"), equalTo(true))
    assertThat(result("(]"), equalTo(false))
    assertThat(result("{[]}"), equalTo(true))
    assertThat(result("([)]"), equalTo(false))
}

class ValidParentheses {
    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        fun isValid(s: String): Boolean {
            val stack = ArrayDeque<Char>()
            for (c in s) {
                when (c) {
                    '(', '[', '{' -> stack.push(c)
                    ')' -> if (stack.isEmpty() || stack.pop() != '(') return false
                    ']' -> if (stack.isEmpty() || stack.pop() != '[') return false
                    '}' -> if (stack.isEmpty() || stack.pop() != '{') return false
                }
            }
            return stack.isEmpty()
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}